src/Shared/TagWriter/TagWriter.cs

```diff
@@ -105,47 +111,57 @@ internal abstract class TagWriter<TTagState, TArrayState>
 
-    protected abstract void WriteStringTag(ref TTagState state, string key, string value);
+    protected abstract void WriteStringTag(ref TTagState state, string key, ReadOnlySpan<char> value);
 
     protected abstract void WriteArrayTag(ref TTagState state, string key, ref TArrayState value);
 
     protected abstract void OnUnsupportedTagDropped(
         string tagKey,
         string tagValueTypeFullName);
 
-    [return: NotNullIfNotNull(nameof(value))]
-    private static string? TruncateString(string? value, int? maxLength)
+    private static ReadOnlySpan<char> TruncateString(ReadOnlySpan<char> value, int? maxLength)
     {
-        return maxLength.HasValue && value?.Length > maxLength
-            ? value.Substring(0, maxLength.Value)
+        return maxLength.HasValue && value.Length > maxLength
+            ? value.Slice(0, maxLength.Value)
             : value;
     }
 
+    private void WriteCharTag(ref TTagState state, string key, char value)
+    {
+        Span<char> destination = stackalloc char[1];
+        destination[0] = value;
+        this.WriteStringTag(ref state, key, destination);
+    }
+
+    private void WriteCharValue(ref TArrayState state, char value)
+    {
+        Span<char> destination = stackalloc char[1];
+        destination[0] = value;
+        this.arrayWriter.WriteStringValue(ref state, destination);
+    }
+
     private void WriteArrayTagInternal(ref TTagState state, string key, Array array, int? tagValueMaxLength)
     {
         var arrayState = this.arrayWriter.BeginWriteArray();
 
         // This switch ensures the values of the resultant array-valued tag are of the same type.
         switch (array)
         {
-            case char[] charArray: this.WriteToArray(ref arrayState, charArray); break;
-            case string[]: this.ConvertToStringArrayThenWriteArrayTag(ref arrayState, array, tagValueMaxLength); break;
-            case bool[] boolArray: this.WriteToArray(ref arrayState, boolArray); break;
-            case byte[] byteArray: this.WriteToArray(ref arrayState, byteArray); break;
-            case sbyte[] sbyteArray: this.WriteToArray(ref arrayState, sbyteArray); break;
-            case short[] shortArray: this.WriteToArray(ref arrayState, shortArray); break;
-            case ushort[] ushortArray: this.WriteToArray(ref arrayState, ushortArray); break;
-            case uint[] uintArray: this.WriteToArray(ref arrayState, uintArray); break;
+            case char[] charArray: this.WriteStructToArray(ref arrayState, charArray); break;
+            case string?[] stringArray: this.WriteStringsToArray(ref arrayState, stringArray, tagValueMaxLength); break;
+            case bool[] boolArray: this.WriteStructToArray(ref arrayState, boolArray); break;
+            case byte[] byteArray: this.WriteToArrayCovariant(ref arrayState, byteArray); break;
+            case short[] shortArray: this.WriteToArrayCovariant(ref arrayState, shortArray); break;
```

Comment:
Why some of the entries like `ushortArray` checks were removed. Are these types not expected here?

Reply:
Looks like it is covered with `WriteToArrayTypeChecked`. Is there a benefit of splitting it out?

Reply:
Kind of an interesting situation. Check out this fiddle: https://dotnetfiddle.net/xLeGFw

Basically runtime treats sbyte[]/byte[], int[]/uint[], short[]/ushort[], and long[]/ulong[] as the same types when it comes to `is`. The way the code was written before we wouldn't hit all the branches, whichever was first for one of those pairs would win and we would end up casting wrong. Like sbyte[] falls into byte[] which means values like -128 ((sbyte)0x80) becomes 128 ((byte)0x80).